aya-build/src/lib.rs

```diff
@@ -202,16 +203,45 @@ impl<'a> Toolchain<'a> {
 
 /// Emit cfg flags that describe the desired BPF target architecture.
 pub fn emit_bpf_target_arch_cfg() {
-    const RUSTFLAGS: &str = "CARGO_ENCODED_RUSTFLAGS";
-
-    println!("cargo:rerun-if-env-changed={RUSTFLAGS}");
-    let rustc_cfgs = std::env::var_os(RUSTFLAGS).unwrap_or_else(|| panic!("{RUSTFLAGS} not set"));
-    let rustc_cfgs = rustc_cfgs
-        .to_str()
-        .unwrap_or_else(|| panic!("{RUSTFLAGS}={rustc_cfgs:?} not unicode"));
-    if !rustc_cfgs.contains("bpf_target_arch") {
-        let arch = target_arch();
-        println!("cargo:rustc-cfg=bpf_target_arch=\"{arch}\"");
+    // The presence of this environment variable indicates that `--cfg
+    // bpf_target_arch="..."` was passed to the compiler, so we don't need to
+    // emit it again. Note that we cannot *set* this environment variable - it
+    // is set by cargo.
+    const BPF_TARGET_ARCH: &str = "CARGO_CFG_BPF_TARGET_ARCH";
+    println!("cargo:rerun-if-env-changed={BPF_TARGET_ARCH}");
+
+    // Users may directly set this environment variable in situations where
+    // using RUSTFLAGS to set `--cfg bpf_target_arch="..."` is not possible or
+    // not ergonomic. In contrast to RUSTFLAGS this mechanism reuses the target
+    // cache for all values, producing many more invalidations.
+    const AYA_BPF_TARGET_ARCH: &str = "AYA_BPF_TARGET_ARCH";
+    println!("cargo:rerun-if-env-changed={AYA_BPF_TARGET_ARCH}");
+
+    const HOST: &str = "HOST";
+    println!("cargo:rerun-if-env-changed={HOST}");
+
+    if std::env::var_os(BPF_TARGET_ARCH).is_none() {
+        let host = std::env::var_os(HOST).unwrap_or_else(|| panic!("{HOST} not set"));
```

Comment:
shouldn't we check CARGO_CFG_TARGET_ARCH before HOST for the cross compilation case? If I'm compiling on x86_64 for aarch64, I need bpf_target_arch=aarch64 or things won't work

Reply:
we just aren't checking `CARGO_CFG_TARGET_ARCH` at all here anymore. If you're cross compiling you are gonna set `AYA_BPF_TARGET_ARCH` because `CARGO_CFG_TARGET_ARCH` is going to be "bpfel".